### PR TITLE
initial usermanager cli

### LINKED ISSUE #3233: Add CLI for user managment
It would be great to have a CLI to manage users in the same fashion as it's now possible with extensions. This CLI should work for all backends. That way, you can create users automatically using automation tools such as ansible or your own scripts to create users across numerous services.

The nextcloud CLI could serve as an example of possible commands

```
 user
  user:add                            adds a user
  user:delete                         deletes the specified user
  user:disable                        disables the specified user
  user:enable                         enables the specified user
  user:info                           show user info
  user:list                           list configured users
  user:resetpassword                  Resets the password of the named user
  user:setting                        Read and modify user settings
```

### PR DESCRIPTION
**Pull request by kossmac**
Fixes #3233

### CODE REVIEW

**Review by splitbrain — COMMENTED** [review]
Looks good already. I added a few smaller nitpicks.

Would be good to have additional commands to add/remove existing users to/from groups.

File: lib/plugins/usermanager/cli.php [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            }
+        } else {
+            if (!empty($pass)) {
+                $this->error($this->getLang('add_fail'));
+                $this->error($this->getLang('addUser_error_modPass_disabled'));
+                return 1;
+            }
+        }
+
+        if (!$auth->triggerUserMod('create', array($user, $pass, $name, $mail, $grps))) {
```
Comment (splitbrain) [inline lib/plugins/usermanager/cli.php:171]: Please add this to get additional info when something fails.

```
            // plugins may signal errors via msg()
            if(isset($MSG)) foreach($MSG as $msg) {
                if($msg['lvl'] === 'error') $this->error($msg['msg']);
            }
```

File: lib/plugins/usermanager/cli.php [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+        // add to group
+        $options->registerCommand('addtogroup', 'Add user to group(s)');
+        $options->registerArgument('name', 'Username', true, 'addtogroup');
+        $options->registerArgument('group', 'Group(s), comma-seperated', true, 'addtogroup');
+
+        // remove from group
+        $options->registerCommand('removefromgroup', 'Remove user from group(s)');
+        $options->registerArgument('name', 'Username', true, 'removefromgroup');
+        $options->registerArgument('group', 'Group(s), comma-seperated', true, 'removefromgroup');
```
Comment (splitbrain) [inline lib/plugins/usermanager/cli.php:56]: typo: separated

File: lib/plugins/usermanager/cli.php [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    protected function cmdAdd(bool $notify, array $args)
+    {
+        global $auth;
+
+        if (!$auth->canDo('addUser')) {
+            $this->error($this->getLang('nosupport'));
+            return 1;
+        }
+
+        list($user, $name, $mail, $grps, $pass) = $args;
```
Comment (splitbrain) [inline lib/plugins/usermanager/cli.php:150]: The help text says, the order is `add <OPTIONS> <name> <mail> [<full_name>] [<groups>] [<password>]`. And indeed creating a user now puts the email address into the name field.

File: lib/plugins/usermanager/cli.php [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            if ($auth->triggerUserMod('modify', array($name, $changes))) {
+                $this->info($this->getLang('update_ok'));
+            } else {
+                $this->error($this->getLang('update_fail'));
+                return 1;
+            }
+        }
+
+        return 0;
+    }
```
Comment (splitbrain) [inline lib/plugins/usermanager/cli.php:278]: I added a user to `foobar`, then to `foobar2`. Then removing him from `foobar` doesn't work. I'm not sure why. It neither gives an error nor a success message.

**Review by splitbrain — COMMENTED** [review]
Nearly there :-)

**Comment by Klap-in** [comment]
Is it possible to add unit tests for client commands? I guess the in and output of a command should be quite clear..

**Comment by splitbrain** [comment]
I fixed the remaining issues and merged this. Doc updates and unit tests would still be very welcome.

**Pull request merged**